File: src/RegexLearningGuide/Levels/Level5_Expert.cs
using System.Text.RegularExpressions;
using static RegexLearningGuide.Helpers.DemoHelper;

namespace RegexLearningGuide.Levels;

/// <summary>
/// LEVEL 5: EXPERT - Named Groups, Balancing Groups, and Compilation
///
/// What you'll learn:
///   - Named capturing groups: (?&lt;name&gt;pattern)
///   - Balancing groups (unique to .NET): (?&lt;name&gt;-&lt;other&gt;)
///   - Compiled regex: RegexOptions.Compiled
///   - Source-generated regex: [GeneratedRegex] (.NET 7+)
///   - MatchEvaluator delegates for advanced replacements
///   - Regex timeout for safety
/// </summary>
public static class Level5_Expert
{
    public static void Run()
    {
        // ────────────────────────────────────────────────────────────────
        // 5.1  NAMED CAPTURING GROUPS
        // ────────────────────────────────────────────────────────────────

        Section("5.1 Named Capturing Groups — (?<name>pattern)");

        Note("Named groups make patterns more readable and maintainable.");
        Note("Access by name: match.Groups[\"name\"].Value");

        ShowExample(
            "Parse a date with named groups",
            pattern: @"(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})",
            input: "Today is 2024-12-25");

        // Demonstrate accessing named groups programmatically
        Console.WriteLine("  ▸ Accessing named groups in code:");
        var dateMatch = Regex.Match("2024-12-25", @"(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})");
        if (dateMatch.Success)
        {
            Console.WriteLine($"    Year  = {dateMatch.Groups["year"].Value}");
            Console.WriteLine($"    Month = {dateMatch.Groups["month"].Value}");
            Console.WriteLine($"    Day   = {dateMatch.Groups["day"].Value}");
        }
        Console.WriteLine();

        // Named backreference
        ShowExample(
            @"Named backreference \k<word> — Find duplicate words",
            pattern: @"\b(?<word>\w+)\s+\k<word>\b",
            input: "This is is a repeated repeated test");

        // Named groups in replacements
        ShowReplace(
            "Use ${name} in replacement strings",
            pattern: @"(?<first>\w+)\s+(?<last>\w+)",
            input: "John Smith",
            replacement: "${last}, ${first}");

        // ────────────────────────────────────────────────────────────────
        // 5.2  BALANCING GROUPS — UNIQUE TO .NET
        // ────────────────────────────────────────────────────────────────

        Section("5.2 Balancing Groups — Match nested structures (unique to .NET!)");

        Note("Balancing groups can match properly nested structures like parentheses.");
        Note("Syntax: (?<open>\\() pushes, (?<-open>\\)) pops from the stack.");
        Note("(?(open)(?!)) at the end asserts the stack is empty (balanced).");

        // Pattern explanation:
        // ^                  — start of string
        // (                  — outer group for the whole expression
        //   [^()]*           — match non-paren characters
        //   (                — start group for nested structure
        //     (              — alternation group
        //       (?<open>\()  — on '(', push to 'open' stack
        //       |
        //       (?<-open>\)) — on ')', pop from 'open' stack (fails if empty)
        //     )
        //     [^()]*         — match non-paren characters
        //   )+               — repeat for nested parens
        // )
        // (?(open)(?!))      — assert 'open' stack is empty
        // $                  — end of string

        string balancingPattern = @"^[^()]*(((?<open>\()|(?<-open>\)))[^()]*)+$";

        Console.WriteLine("  ▸ Matching balanced parentheses:");
        Console.WriteLine($"    Pattern: {balancingPattern}");
        Console.WriteLine();

        string[] testStrings = [
            "(a + b)",
            "((a + b) * (c + d))",
            "(a + (b * c)",      // unbalanced
            "a + b) * c(",       // unbalanced
            "((()))",
        ];

        foreach (var test in testStrings)
        {
            // Use a simpler but effective balancing group pattern
            string pattern = @"^\((?:[^()]|(?<open>\()|(?<-open>\)))*(?(open)(?!))\)$";
            bool isBalanced = Regex.IsMatch(test, pattern);
            Console.WriteLine($"    \"{test}\" => {(isBalanced ? "BALANCED" : "UNBALANCED")}");
        }
        Console.WriteLine();

        // ────────────────────────────────────────────────────────────────
        // 5.3  MATCHEVALUATOR — DYNAMIC REPLACEMENTS
        // ────────────────────────────────────────────────────────────────

        Section("5.3 MatchEvaluator — Use delegates for dynamic replacements");

        // Replace with a delegate that transforms each match
        Console.WriteLine("  ▸ Double every number in a string:");
        string input1 = "I have 3 cats and 5 dogs and 12 fish";
        string result1 = Regex.Replace(input1, @"\d+", match =>
        {
            int number = int.Parse(match.Value);
            return (number * 2).ToString();
        });
        Console.WriteLine($"    Input  : \"{input1}\"");
        Console.WriteLine($"    Result : \"{result1}\"");
        Console.WriteLine();

        // Title case conversion
        Console.WriteLine("  ▸ Convert to Title Case:");
        string input2 = "the quick brown fox jumps over the lazy dog";
        string result2 = Regex.Replace(input2, @"\b\w", match =>
            match.Value.ToUpper());
        Console.WriteLine($"    Input  : \"{input2}\"");
        Console.WriteLine($"    Result : \"{result2}\"");
        Console.WriteLine();

        // Mask sensitive data
        Console.WriteLine("  ▸ Mask credit card numbers:");
        string input3 = "Card: 4111-1111-1111-1111, Alt: 5500-0000-0000-0004";
        string result3 = Regex.Replace(input3, @"\d{4}-\d{4}-\d{4}-(\d{4})", match =>
            $"****-****-****-{match.Groups[1].Value}");
        Console.WriteLine($"    Input  : \"{input3}\"");
        Console.WriteLine($"    Result : \"{result3}\"");
        Console.WriteLine();

        // ────────────────────────────────────────────────────────────────
        // 5.4  COMPILED REGEX & SOURCE GENERATORS
        // ────────────────────────────────────────────────────────────────

        Section("5.4 Compiled Regex & Source Generators");

        Note("RegexOptions.Compiled — Compiles pattern to IL for faster execution.");
        Note("Best for patterns used many times. Slower startup, faster matching.");

        Console.WriteLine("  ▸ Creating a compiled regex:");
        Console.WriteLine("    var regex = new Regex(@\"\\d+\", RegexOptions.Compiled);");
        Console.WriteLine();

        // Demonstrate compiled regex
        var compiledRegex = new Regex(@"\d+", RegexOptions.Compiled);
        var match2 = compiledRegex.Match("Order #12345");
        Console.WriteLine($"    Match: \"{match2.Value}\" at index {match2.Index}");
        Console.WriteLine();

        Note(".NET 7+ introduced [GeneratedRegex] for compile-time source generation:");
        Console.WriteLine("    [GeneratedRegex(@\"\\d+\")]");
        Console.WriteLine("    private static partial Regex NumberPattern();");
        Console.WriteLine();
        Note("Source generators produce C# code at compile time — zero runtime cost!");

        // Show the source-generated regex
        var genMatch = NumberPattern().Match("Order #12345");
        Console.WriteLine($"  ▸ Source-generated regex match: \"{genMatch.Value}\"");
        Console.WriteLine();

        // ────────────────────────────────────────────────────────────────
        // 5.5  REGEX TIMEOUT — PREVENT CATASTROPHIC BACKTRACKING
        // ────────────────────────────────────────────────────────────────

        Section("5.5 Regex Timeout — Safety against catastrophic backtracking");

        Note("Some patterns can take exponential time on certain inputs.");
        Note("Always set a timeout when processing untrusted input!");

        Console.WriteLine("  ▸ Creating a regex with a 2-second timeout:");
        Console.WriteLine("    var regex = new Regex(pattern, RegexOptions.None,");
        Console.WriteLine("                         TimeSpan.FromSeconds(2));");
        Console.WriteLine();

        // Demonstrate timeout
        try
        {
            // This pattern is intentionally crafted to demonstrate timeout.
            // (a+)+ is a classic catastrophic backtracking pattern.
            var riskyRegex = new Regex(@"^(a+)+$", RegexOptions.None,
                TimeSpan.FromMilliseconds(500));
            // Short input works fine
            Console.WriteLine($"    Short input 'aaaaab': IsMatch = {riskyRegex.IsMatch("aaaaab")}");

            // Long input might timeout
            string longInput = new string('a', 25) + "b";
            Console.WriteLine($"    Long input (25 a's + b): trying with 500ms timeout...");
            bool result = riskyRegex.IsMatch(longInput);
            Console.WriteLine($"    Result: {result}");
        }
        catch (RegexMatchTimeoutException ex)
        {
            Console.WriteLine($"    TIMEOUT! Regex exceeded {ex.MatchTimeout.TotalMilliseconds}ms");
            Console.WriteLine("    This is why timeouts are important for untrusted input!");
        }
        Console.WriteLine();
    }

    // Source-generated regex (compile-time optimization, .NET 7+)
    [GeneratedRegex(@"\d+")]
    private static partial Regex NumberPattern();
}

File: ts-regex-guide/src/levels/level5-expert.ts
import { section, note, showExample, showReplace } from "../helpers/demo-helper.js";

// ============================================================================
// LEVEL 5: EXPERT — Named Groups, Unicode, and Advanced Techniques
//
// What you'll learn:
//   - Named capturing groups: (?<name>pattern)
//   - Named backreferences: \k<name>
//   - Unicode support: /u flag, \p{} Unicode property escapes
//   - Unicode sets: /v flag (ES2024)
//   - Replace with function callbacks
//   - String.raw for readable patterns
//   - Recursive-style matching techniques
// ============================================================================

export function run(): void {
  // ──────────────────────────────────────────────────────────────────
  // 5.1  NAMED CAPTURING GROUPS
  // ──────────────────────────────────────────────────────────────────
  section("5.1 Named Capturing Groups — (?<name>pattern)");

  note("Named groups make patterns more readable and maintainable.");
  note("Access by name: match.groups.name or match.groups['name']");

  showExample(
    "Parse a date with named groups",
    /(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})/g,
    "Today is 2024-12-25"
  );

  // Demonstrate accessing named groups programmatically
  console.log("  ▸ Accessing named groups in code:");
  const dateMatch = "2024-12-25".match(/(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})/);
  if (dateMatch?.groups) {
    console.log(`    Year  = ${dateMatch.groups.year}`);
    console.log(`    Month = ${dateMatch.groups.month}`);
    console.log(`    Day   = ${dateMatch.groups.day}`);
  }
  console.log();

  // Destructuring named groups
  console.log("  ▸ Destructuring named groups (clean syntax):");
  const { groups: { year, month, day } = {} } =
    "2024-12-25".match(/(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})/) ?? {};
  console.log(`    const { groups: { year, month, day } } = "2024-12-25".match(pattern);`);
  console.log(`    year=${year}, month=${month}, day=${day}`);
  console.log();

  // Named backreference
  showExample(
    "Named backreference \\k<word> — Find duplicate words",
    /\b(?<word>\w+)\s+\k<word>\b/g,
    "This is is a repeated repeated test"
  );

  // Named groups in replacements
  showReplace(
    "Use $<name> in replacement strings",
    /(?<first>\w+)\s+(?<last>\w+)/,
    "John Smith",
    "$<last>, $<first>"
  );

  // ──────────────────────────────────────────────────────────────────
  // 5.2  UNICODE SUPPORT — /u FLAG
  // ──────────────────────────────────────────────────────────────────
  section("5.2 Unicode Support — The /u Flag");

  note("The 'u' flag enables full Unicode matching and \\p{} property escapes.");

  // Without /u, emoji and astral plane characters don't work correctly
  console.log("  ▸ Without /u flag — emoji is treated as two code units:");
  console.log(`    "😀".match(/./) = "${("😀".match(/./) ?? [""])[0]}" (broken — partial match)`);
  console.log(`    "😀".length = ${"😀".length} (two UTF-16 code units)`);
  console.log();

  console.log("  ▸ With /u flag — emoji is treated as one character:");
  console.log(`    "😀".match(/./u) = "${("😀".match(/./u) ?? [""])[0]}" (correct)`);
  console.log();

  // Unicode property escapes
  section("5.2b Unicode Property Escapes — \\p{...} (requires /u flag)");

  showExample(
    "\\p{Letter} — Match any Unicode letter",
    /\p{Letter}+/gu,
    "Hello Мир 世界 مرحبا"
  );

  showExample(
    "\\p{Script=Greek} — Match Greek characters",
    /\p{Script=Greek}+/gu,
    "Hello Ωmega αβγ world"
  );

  showExample(
    "\\p{Emoji} — Match emoji characters",
    /\p{Emoji}/gu,
    "Hello 😀 World 🌍 TypeScript 🚀"
  );

  showExample(
    "\\p{Number} — Match any Unicode number",
    /\p{Number}+/gu,
    "Latin 123 Arabic ١٢٣ Roman Ⅳ"
  );

  showExample(
    "\\P{ASCII} — Match non-ASCII characters (uppercase P = negation)",
    /\P{ASCII}+/gu,
    "Hello café résumé naïve"
  );

  note("Common Unicode categories: Letter, Number, Punctuation, Symbol, Separator, Mark");

  // ──────────────────────────────────────────────────────────────────
  // 5.3  UNICODE SETS — /v FLAG (ES2024)
  // ──────────────────────────────────────────────────────────────────
  section("5.3 Unicode Sets — The /v Flag (ES2024)");

  note("The 'v' flag is an upgrade of 'u' — enables set operations inside [...].");
  note("Supports: intersection (&&), subtraction (--), nested classes.");

  // Set subtraction: match letters but not vowels
  showExample(
    "[\\p{Letter}--[aeiouAEIOU]] — Letters minus vowels",
    /[\p{Letter}--[aeiouAEIOU]]+/gv,
    "Hello World"
  );

  // Set intersection: match characters that are both ASCII and digits
  showExample(
    "[\\p{ASCII}&&\\p{Number}] — ASCII digits only",
    /[\p{ASCII}&&\p{Number}]+/gv,
    "Latin 123 Arabic ١٢٣"
  );

  // String literals in character classes (v flag feature)
  showExample(
    "Match specific multi-character sequences in [...]",
    /[\\q{abc|def|ghi}]/gv,
    "a b c d e f abc def ghi"
  );

  // ──────────────────────────────────────────────────────────────────
  // 5.4  REPLACE WITH FUNCTION CALLBACKS
  // ──────────────────────────────────────────────────────────────────
  section("5.4 Replace with Function Callbacks — Dynamic replacements");

  // Double every number
  console.log("  ▸ Double every number in a string:");
  const input1 = "I have 3 cats and 5 dogs and 12 fish";
  const result1 = input1.replace(/\d+/g, (match) => {
    return (parseInt(match) * 2).toString();
  });
  console.log(`    Input  : "${input1}"`);
  console.log(`    Result : "${result1}"`);
  console.log();

  // Title case
  console.log("  ▸ Convert to Title Case:");
  const input2 = "the quick brown fox jumps over the lazy dog";
  const result2 = input2.replace(/\b\w/g, (match) => match.toUpperCase());
  console.log(`    Input  : "${input2}"`);
  console.log(`    Result : "${result2}"`);
  console.log();

  // Callback with groups
  console.log("  ▸ Replace callback with named groups:");
  const input3 = "2024-12-25 and 2024-01-01";
  const result3 = input3.replace(
    /(?<y>\d{4})-(?<m>\d{2})-(?<d>\d{2})/g,
    (_match, _p1, _p2, _p3, _offset, _str, groups) => {
      return `${groups.m}/${groups.d}/${groups.y}`;
    }
  );
  console.log(`    Input  : "${input3}"`);
  console.log(`    Result : "${result3}"`);
  console.log();

  note("Callback signature: (match, p1, p2, ..., offset, string, groups) => replacement");

  // Mask sensitive data
  console.log("  ▸ Mask credit card numbers:");
  const input4 = "Card: 4111-1111-1111-1111, Alt: 5500-0000-0000-0004";
  const result4 = input4.replace(
    /\d{4}-\d{4}-\d{4}-(\d{4})/g,
    (_match, last4) => `****-****-****-${last4}`
  );
  console.log(`    Input  : "${input4}"`);
  console.log(`    Result : "${result4}"`);
  console.log();

  // ──────────────────────────────────────────────────────────────────
  // 5.5  STRING.RAW FOR READABLE PATTERNS
  // ──────────────────────────────────────────────────────────────────
  section("5.5 String.raw + new RegExp() for readable patterns");

  note("Use String.raw to avoid double-escaping in new RegExp() constructor.");

  console.log("  ▸ Without String.raw (must double-escape):");
  console.log('    new RegExp("\\\\d+\\\\.\\\\d+")  // messy');
  console.log();

  console.log("  ▸ With String.raw (single escape, like literal syntax):");
  console.log('    new RegExp(String.raw`\\d+\\.\\d+`)  // clean');
  console.log();

  const rawRegex = new RegExp(String.raw`\d+\.\d+`, "g");
  const rawResult = "Price: 19.99 and 5.50".match(rawRegex);
  console.log(`    Result: [${rawResult?.map((r) => `"${r}"`).join(", ")}]`);
  console.log();

  // Building patterns dynamically
  console.log("  ▸ Building patterns dynamically with template literals:");
  const allowedExtensions = ["jpg", "png", "gif", "webp"];
  const extPattern = new RegExp(
    String.raw`\.(?:${allowedExtensions.join("|")})$`,
    "i"
  );
  console.log(`    Pattern: ${extPattern}`);
  console.log(`    "photo.jpg" : ${extPattern.test("photo.jpg")}`);
  console.log(`    "doc.pdf"   : ${extPattern.test("doc.pdf")}`);
  console.log();

  // ──────────────────────────────────────────────────────────────────
  // 5.6  REGEX AND ITERATORS — Symbol.matchAll, Symbol.replace, etc.
  // ──────────────────────────────────────────────────────────────────
  section("5.6 Well-Known Symbols — Customize regex behavior");

  note("JavaScript uses Symbol.match, Symbol.replace, Symbol.search, Symbol.split");
  note("to allow objects to behave like regex in string methods.");

  // Custom matcher object
  console.log("  ▸ Custom matcher using Symbol.match:");

  const vowelMatcher = {
    [Symbol.match](str: string) {
      return str.match(/[aeiou]/gi);
    },
  };

  // @ts-expect-error — custom Symbol.match object
  const vowels = "Hello World".match(vowelMatcher);
  console.log(`    "Hello World".match(vowelMatcher) = [${vowels?.join(", ")}]`);
  console.log();

  // Custom replacer
  console.log("  ▸ Custom replacer using Symbol.replace:");

  const censor = {
    [Symbol.replace](str: string, replacement: string) {
      return str.replace(/[aeiou]/gi, replacement);
    },
  };

  // @ts-expect-error — custom Symbol.replace object
  const censored = "Hello World".replace(censor, "*");
  console.log(`    "Hello World".replace(censor, "*") = "${censored}"`);
  console.log();
}
